File: app/models/schema.py
import warnings
from enum import Enum
from typing import Any, List, Literal, Optional, Union

import pydantic
from pydantic import BaseModel, ConfigDict, Field

from app.config import config

# Pydantic 의 특정 경고를 무시한다
warnings.filterwarnings(
    "ignore",
    category=UserWarning,
    message="Field name.*shadows an attribute in parent.*",
)


class VideoConcatMode(str, Enum):
    random = "random"
    sequential = "sequential"


class VideoTransitionMode(str, Enum):
    none = None
    shuffle = "Shuffle"
    fade_in = "FadeIn"
    fade_out = "FadeOut"
    slide_in = "SlideIn"
    slide_out = "SlideOut"
    zoom_in = "ZoomIn"
    zoom_out = "ZoomOut"


class VideoAspect(str, Enum):
    landscape = "16:9"
    portrait = "9:16"
    square = "1:1"

    def to_resolution(self):
        if self == VideoAspect.landscape:
            return 1920, 1080
        elif self == VideoAspect.portrait:
            return 1080, 1920
        elif self == VideoAspect.square:
            return 1080, 1080
        raise ValueError(f"unsupported video aspect: {self}")


class _Config:
    arbitrary_types_allowed = True


@pydantic.dataclasses.dataclass(config=_Config)
class MaterialInfo:
    provider: str = "pexels"
    url: str = ""
    duration: int = 0
    # 온라인 소재 검색은 걸러진 공개 출처 정보를 함께 담아 검색 캐시와 작업 기록에서
    # 재사용한다. 로컬 업로드 소재는 채울 필요가 없다. 작업 파일에 쓰기 전에 필드
    # 화이트리스트로 다시 구성하므로, 외부 요청이 넘긴 서명 URL, 자격 증명, 무관한
    # 필드가 영속 데이터에 들어가지 않는다.
    source_info: Optional[dict[str, Any]] = None


class VideoParams(BaseModel):
    """
    {
      "video_subject": "",
      "video_aspect": "가로 16:9",
      "voice_name": "ko-KR-SunHiNeural-Female",
      "bgm_name": "random",
      "font_name": "Pretendard-Bold.ttf",
      "text_color": "#FFFFFF",
      "font_size": 60,
      "stroke_color": "#000000",
      "stroke_width": 1.5
    }
    """

    video_subject: str
    video_script: str = ""  # Script used to generate the video
    video_terms: Optional[str | list] = None  # Keywords used to generate the video
    video_aspect: Optional[VideoAspect] = VideoAspect.portrait.value
    video_concat_mode: Optional[VideoConcatMode] = VideoConcatMode.random.value
    video_transition_mode: Optional[VideoTransitionMode] = None
    # 숏폼은 2~4 초마다 화면이 바뀌는 편집이 시청 지속에 유리하다. WebUI 기본값도
    # 3 이었는데 스키마만 5 로 남아, CLI 와 API 만 느리게 잘리고 있었다.
    video_clip_duration: Optional[int] = 3
    video_clip_speed: Optional[float] = 1.0
    match_materials_to_script: bool = False
    video_count: Optional[int] = 1

    video_source: Optional[str] = "pexels"
    video_materials: Optional[List[MaterialInfo]] = (
        None  # Materials used to generate the video
    )
    
    custom_audio_file: Optional[str] = None  # Custom audio file path, will ignore TTS and can still use Whisper subtitles
    video_language: Optional[str] = ""  # auto detect

    voice_name: Optional[str] = ""
    voice_volume: Optional[float] = 1.0
    voice_rate: Optional[float] = 1.0
    bgm_type: Optional[str] = "random"
    bgm_file: Optional[str] = ""
    bgm_volume: Optional[float] = 0.2
    # 영상 배경음악 제공자들이 함께 쓰는 프롬프트로, WebUI 의 새 작업은 이 필드에 기록한다.
    # 아래 Sonilo 전용 필드는 예전 작업 기록과 기존 CLI 파라미터 호환을 위해 남겨 둔다.
    video_music_prompt: str = Field(default="", max_length=2000)
    sonilo_bgm_prompt: str = Field(default="", max_length=2000)

    subtitle_enabled: Optional[bool] = True
    subtitle_position: Optional[str] = config.ui.get("subtitle_position", "bottom")  # top, bottom, center, custom
    custom_position: float = config.ui.get("custom_position", 70.0)
    font_name: Optional[str] = "Pretendard-Bold.ttf"
    text_fore_color: Optional[str] = "#FFFFFF"
    text_background_color: Union[bool, str] = False
    rounded_subtitle_background: bool = False

    # 쇼츠 템플릿 레이아웃. "fullscreen" 은 기존 동작(영상이 화면을 꽉 채움)이고,
    # "card" 는 영상을 축소해 배경 캔버스 위에 얹어, 위아래로 헤드라인과 자막을
    # 놓을 여백을 만든다. 유튜브 쇼츠에서 흔한 구성이다.
    layout: Optional[str] = "fullscreen"
    layout_background_color: str = "#FFFFFF"
    # 영상이 차지할 세로 비율. 0.55 면 화면 높이의 55% 를 영상이 쓰고 나머지가 여백이다.
    layout_video_height_ratio: float = Field(default=0.55, ge=0.3, le=1.0)
    # 화면 상단에 얹는 후킹 문구. 비워 두면 card 레이아웃에서 LLM 이 생성한다.
    headline: str = Field(default="", max_length=200)
    headline_color: str = "#111111"
    headline_stroke_color: str = ""
    headline_font_size: int = Field(default=92, ge=20, le=200)
    # card 레이아웃에서 자막을 영상 아래 여백에 놓는다. 영상 위에 얹으면 화면이
    # 가려지고, 여백은 배경색이라 자막 색도 따로 받아야 한다.
    subtitle_below_video: bool = False
    subtitle_below_color: str = "#111111"
    # 영상 모서리를 둥글게 깎는다. 0 이면 각진 그대로.
    layout_corner_radius: int = Field(default=0, ge=0, le=120)

    font_size: int = 60
    stroke_color: Optional[str] = "#000000"
    stroke_width: float = 1.5
    n_threads: Optional[int] = 2
    paragraph_number: int = Field(default=1, ge=1, le=10)
    video_script_prompt: str = Field(default="", max_length=2000)
    custom_system_prompt: str = Field(default="", max_length=8000)


class SubtitleRequest(BaseModel):
    video_script: str
    video_language: Optional[str] = ""
    voice_name: Optional[str] = "zh-CN-XiaoxiaoNeural-Female"
    voice_volume: Optional[float] = 1.0
    voice_rate: Optional[float] = 1.2
    bgm_type: Optional[str] = "random"
    bgm_file: Optional[str] = ""
    bgm_volume: Optional[float] = 0.2
    subtitle_position: Optional[str] = config.ui.get("subtitle_position", "bottom")
    font_name: Optional[str] = "Pretendard-Bold.ttf"
    text_fore_color: Optional[str] = "#FFFFFF"
    text_background_color: Union[bool, str] = False
    rounded_subtitle_background: bool = False
    font_size: int = 60
    stroke_color: Optional[str] = "#000000"
    stroke_width: float = 1.5
    video_source: Optional[str] = "local"
    subtitle_enabled: Optional[str] = "true"


class AudioRequest(BaseModel):
    video_script: str
    video_language: Optional[str] = ""
    voice_name: Optional[str] = "zh-CN-XiaoxiaoNeural-Female"
    voice_volume: Optional[float] = 1.0
    voice_rate: Optional[float] = 1.2
    bgm_type: Optional[str] = "random"
    bgm_file: Optional[str] = ""
    bgm_volume: Optional[float] = 0.2
    video_source: Optional[str] = "local"


class VideoScriptParams:
    """
    {
      "video_subject": "봄날의 꽃바다",
      "video_language": "",
      "paragraph_number": 1,
      "video_script_prompt": "",
      "custom_system_prompt": ""
    }
    """

    video_subject: Optional[str] = "봄날의 꽃바다"
    video_language: Optional[str] = ""
    paragraph_number: int = Field(default=1, ge=1, le=10)
    video_script_prompt: str = Field(default="", max_length=2000)
    custom_system_prompt: str = Field(default="", max_length=8000)


class VideoTermsParams:
    """
    {
      "video_subject": "",
      "video_script": "",
      "amount": 5,
      "match_materials_to_script": false
    }
    """

    video_subject: Optional[str] = "봄날의 꽃바다"
    video_script: Optional[str] = (
        "봄날의 꽃바다가 한 폭의 그림처럼 눈앞에 펼쳐집니다. 만물이 깨어나는 계절, 대지는 화려한 색의 옷을 갈아입습니다. 노란 개나리, 연분홍 벚꽃, 새하얀 배꽃, 선명한 튤립까지……"
    )
    amount: Optional[int] = 5
    match_materials_to_script: bool = False


class VideoSocialMetadataParams:
    """
    {
      "video_subject": "A day in Shanghai",
      "video_script": "",
      "language": "auto",
      "platform": "tiktok"
    }
    """

    video_subject: Optional[str] = Field(default="A day in Shanghai", max_length=500)
    video_script: Optional[str] = Field(default="", max_length=8000)
    language: Optional[str] = Field(default="auto", max_length=64)
    platform: Optional[str] = Field(default="tiktok", max_length=64)


class BaseResponse(BaseModel):
    status: int = 200
    message: Optional[str] = "success"
    data: Any = None


class TaskVideoRequest(VideoParams, BaseModel):
    pass


class TaskQueryRequest(BaseModel):
    pass


class VideoScriptRequest(VideoScriptParams, BaseModel):
    pass


class VideoTermsRequest(VideoTermsParams, BaseModel):
    pass


class VideoSocialMetadataRequest(VideoSocialMetadataParams, BaseModel):
    pass


######################################################################################################
######################################################################################################
######################################################################################################
######################################################################################################
class TaskResponse(BaseResponse):
    class TaskResponseData(BaseModel):
        task_id: str

    data: TaskResponseData

    class Config:
        json_schema_extra = {
            "example": {
                "status": 200,
                "message": "success",
                "data": {"task_id": "6c85c8cc-a77a-42b9-bc30-947815aa0558"},
            },
        }


class TaskStatusData(BaseModel):
    """작업 조회가 외부에 보장하는 안정 필드. 예전 필드와 확장 필드는 그대로 통과시킨다."""

    model_config = ConfigDict(extra="allow")

    task_id: str
    state: int
    progress: int = 0
    videos: Optional[List[str]] = None
    combined_videos: Optional[List[str]] = None
    failed_stage: Optional[str] = None
    error: Optional[str] = None
    cross_post_state: Optional[
        Literal["pending", "processing", "complete", "failed"]
    ] = None
    cross_post_results: Optional[List[dict[str, Any]]] = None
    cross_post_error: Optional[str] = None


class TaskListData(BaseModel):
    """페이지네이션된 작업 목록 구조."""

    tasks: List[TaskStatusData]
    total: int
    page: int
    page_size: int


class TaskQueryResponse(BaseResponse):
    """
    작업 조회는 생성 상태와, 선택적으로 플랫폼 업로드 상태를 반환한다.

    생성이 실패하면 `failed_stage` 와 `error` 가 포함된다. 생성이 끝난 뒤 자동 업로드가
    켜져 있으면 `cross_post_state` 가 pending, processing, complete, failed 순으로 진행된다.
    """

    data: TaskStatusData

    model_config = ConfigDict(
        json_schema_extra={
            "examples": [
                {
                    "status": 200,
                    "message": "success",
                    "data": {
                        "task_id": "6c85c8cc-a77a-42b9-bc30-947815aa0558",
                        "state": 1,
                        "progress": 100,
                        "videos": ["/tasks/example/final-1.mp4"],
                        "cross_post_state": "complete",
                        "cross_post_results": [{"success": True}],
                    },
                },
                {
                    "status": 200,
                    "message": "success",
                    "data": {
                        "task_id": "6c85c8cc-a77a-42b9-bc30-947815aa0558",
                        "state": -1,
                        "progress": 30,
                        "failed_stage": "audio",
                        "error": "TTS request timed out",
                    },
                },
            ],
        }
    )


class TaskListResponse(BaseResponse):
    """작업 목록은 별도 응답 모델을 쓴다. 단일 작업 조회와 문서 구조가 섞이지 않게 하기 위해서다."""

    data: TaskListData

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "status": 200,
                "message": "success",
                "data": {
                    "tasks": [
                        {
                            "task_id": "6c85c8cc-a77a-42b9-bc30-947815aa0558",
                            "state": 4,
                            "progress": 50,
                        }
                    ],
                    "total": 1,
                    "page": 1,
                    "page_size": 10,
                },
            }
        }
    )


class TaskDeletionResponse(BaseResponse):
    class Config:
        json_schema_extra = {
            "example": {
                "status": 200,
                "message": "success",
                "data": {
                    "state": 1,
                    "progress": 100,
                    "videos": [
                        "http://127.0.0.1:8080/tasks/6c85c8cc-a77a-42b9-bc30-947815aa0558/final-1.mp4"
                    ],
                    "combined_videos": [
                        "http://127.0.0.1:8080/tasks/6c85c8cc-a77a-42b9-bc30-947815aa0558/combined-1.mp4"
                    ],
                },
            },
        }


class VideoScriptResponse(BaseResponse):
    class Config:
        json_schema_extra = {
            "example": {
                "status": 200,
                "message": "success",
                "data": {
                    "video_script": "봄날의 꽃바다는 자연이 그린 한 폭의 아름다운 그림입니다. 이 계절이면 대지가 깨어나고 만물이 자라며, 꽃들이 앞다투어 피어나 오색찬란한 꽃바다를 이룹니다..."
                },
            },
        }


class VideoTermsResponse(BaseResponse):
    class Config:
        json_schema_extra = {
            "example": {
                "status": 200,
                "message": "success",
                "data": {"video_terms": ["sky", "tree"]},
            },
        }


class VideoSocialMetadataResponse(BaseResponse):
    class Config:
        json_schema_extra = {
            "example": {
                "status": 200,
                "message": "success",
                "data": {
                    "title": "A Day in Shanghai You Should Not Miss",
                    "caption": "Save this quick Shanghai inspiration and follow for more short travel ideas.",
                    "hashtags": ["#shorts", "#travel", "#shanghai", "#viral", "#fyp"],
                },
            },
        }


class BgmRetrieveResponse(BaseResponse):
    class Config:
        json_schema_extra = {
            "example": {
                "status": 200,
                "message": "success",
                "data": {
                    "files": [
                        {
                            "name": "4fca18fce7344f3aa824777a40d45c8c.mp3",
                            "size": 1891269,
                            "file": "4fca18fce7344f3aa824777a40d45c8c.mp3",
                        }
                    ]
                },
            },
        }


class BgmUploadResponse(BaseResponse):
    class Config:
        json_schema_extra = {
            "example": {
                "status": 200,
                "message": "success",
                "data": {"file": "4fca18fce7344f3aa824777a40d45c8c.mp3"},
            },
        }

class VideoMaterialRetrieveResponse(BaseResponse):
    class Config:
        json_schema_extra = {
            "example": {
                "status": 200,
                "message": "success",
                "data": {
                    "files": [
                        {
                            "name": "example.mp4",
                            "size": 12345678,
                            "file": "/MoneyPrinterTurbo/resource/videos/example.mp4",
                        }
                    ]
                },
            },
        }

class VideoMaterialUploadResponse(BaseResponse):
    class Config:
        json_schema_extra = {
            "example": {
                "status": 200,
                "message": "success",
                "data": {
                    "file": "/MoneyPrinterTurbo/resource/videos/example.mp4",
                },
            },
        }

File: go/internal/models/schema.go
// 请求/任务数据结构，对应 app/models/schema.py。
package models

import (
	"encoding/json"
	"fmt"
	"strings"
)

// VideoConcatMode 素材拼接顺序。
type VideoConcatMode string

const (
	ConcatRandom     VideoConcatMode = "random"
	ConcatSequential VideoConcatMode = "sequential"
)

// VideoAspect 成片画幅。
type VideoAspect string

const (
	AspectLandscape VideoAspect = "16:9"
	AspectPortrait  VideoAspect = "9:16"
	AspectSquare    VideoAspect = "1:1"
)

// Resolution 返回画幅对应的像素宽高。
func (a VideoAspect) Resolution() (int, int, error) {
	switch a {
	case AspectLandscape:
		return 1920, 1080, nil
	case AspectPortrait:
		return 1080, 1920, nil
	case AspectSquare:
		return 1080, 1080, nil
	default:
		return 0, 0, fmt.Errorf("unsupported video aspect: %s", a)
	}
}

// MaterialInfo 单条视频/图片素材。
type MaterialInfo struct {
	Provider   string         `json:"provider"`
	URL        string         `json:"url"`
	Duration   int            `json:"duration"`
	SourceInfo map[string]any `json:"source_info,omitempty"`
}

// StringOrList 兼容 video_terms 既可能是逗号分隔字符串，也可能是字符串数组。
type StringOrList []string

func (s *StringOrList) UnmarshalJSON(data []byte) error {
	if string(data) == "null" {
		*s = nil
		return nil
	}
	var list []string
	if err := json.Unmarshal(data, &list); err == nil {
		*s = list
		return nil
	}
	var raw string
	if err := json.Unmarshal(data, &raw); err != nil {
		return err
	}
	if strings.TrimSpace(raw) == "" {
		*s = nil
		return nil
	}
	parts := strings.FieldsFunc(raw, func(r rune) bool {
		return r == ',' || r == '，'
	})
	out := make([]string, 0, len(parts))
	for _, p := range parts {
		if t := strings.TrimSpace(p); t != "" {
			out = append(out, t)
		}
	}
	*s = out
	return nil
}

// BoolOrString 对应 text_background_color：false 关闭，true 用默认色，字符串为 #RRGGBB。
type BoolOrString struct {
	Enabled bool
	Color   string
	IsColor bool
}

func (b *BoolOrString) UnmarshalJSON(data []byte) error {
	if string(data) == "null" {
		*b = BoolOrString{}
		return nil
	}
	var flag bool
	if err := json.Unmarshal(data, &flag); err == nil {
		b.Enabled = flag
		b.IsColor = false
		return nil
	}
	var color string
	if err := json.Unmarshal(data, &color); err != nil {
		return err
	}
	b.Enabled = color != ""
	b.Color = color
	b.IsColor = color != ""
	return nil
}

func (b BoolOrString) MarshalJSON() ([]byte, error) {
	if b.IsColor {
		return json.Marshal(b.Color)
	}
	return json.Marshal(b.Enabled)
}

// VideoParams 一次视频生成任务的全部参数，对应 Python VideoParams。
type VideoParams struct {
	VideoSubject            string           `json:"video_subject"`
	VideoScript             string           `json:"video_script"`
	VideoTerms              StringOrList     `json:"video_terms"`
	VideoAspect             VideoAspect      `json:"video_aspect"`
	VideoConcatMode         VideoConcatMode  `json:"video_concat_mode"`
	VideoTransitionMode     string           `json:"video_transition_mode"`
	VideoClipDuration       int              `json:"video_clip_duration"`
	VideoClipSpeed          float64          `json:"video_clip_speed"`
	MatchMaterialsToScript  bool             `json:"match_materials_to_script"`
	VideoCount              int              `json:"video_count"`
	VideoSource             string           `json:"video_source"`
	VideoMaterials          []MaterialInfo   `json:"video_materials"`
	CustomAudioFile         string           `json:"custom_audio_file"`
	VideoLanguage           string           `json:"video_language"`
	VoiceName               string           `json:"voice_name"`
	VoiceVolume             float64          `json:"voice_volume"`
	VoiceRate               float64          `json:"voice_rate"`
	BGMType                 string           `json:"bgm_type"`
	BGMFile                 string           `json:"bgm_file"`
	BGMVolume               float64          `json:"bgm_volume"`
	VideoMusicPrompt        string           `json:"video_music_prompt"`
	SoniloBGMPrompt         string           `json:"sonilo_bgm_prompt"`
	SubtitleEnabled         *bool            `json:"subtitle_enabled"`
	SubtitlePosition        string           `json:"subtitle_position"`
	CustomPosition          float64          `json:"custom_position"`
	FontName                string           `json:"font_name"`
	TextForeColor           string           `json:"text_fore_color"`
	TextBackgroundColor     BoolOrString     `json:"text_background_color"`
	RoundedSubtitleBackground bool           `json:"rounded_subtitle_background"`
	FontSize                int              `json:"font_size"`
	StrokeColor             string           `json:"stroke_color"`
	StrokeWidth             float64          `json:"stroke_width"`
	NThreads                int              `json:"n_threads"`
	ParagraphNumber         int              `json:"paragraph_number"`
	VideoScriptPrompt       string           `json:"video_script_prompt"`
	CustomSystemPrompt      string           `json:"custom_system_prompt"`
}

// DefaultVideoParams 返回与 Python 模型一致的默认值。
func DefaultVideoParams() VideoParams {
	enabled := true
	return VideoParams{
		VideoAspect:       AspectPortrait,
		VideoConcatMode:   ConcatRandom,
		VideoClipDuration: 5,
		VideoClipSpeed:    1.0,
		VideoCount:        1,
		VideoSource:       "pexels",
		VoiceName:         "zh-CN-XiaoxiaoNeural-Female",
		VoiceVolume:       1.0,
		VoiceRate:         1.0,
		BGMType:           "random",
		BGMVolume:         0.2,
		SubtitleEnabled:   &enabled,
		SubtitlePosition:  "bottom",
		CustomPosition:    70,
		FontName:          "STHeitiMedium.ttc",
		TextForeColor:     "#FFFFFF",
		FontSize:          60,
		StrokeColor:       "#000000",
		StrokeWidth:       1.5,
		NThreads:          2,
		ParagraphNumber:   1,
	}
}

// SubtitlesOn 字幕默认开启。
func (p VideoParams) SubtitlesOn() bool {
	if p.SubtitleEnabled == nil {
		return true
	}
	return *p.SubtitleEnabled
}

// ScriptRequest 仅生成文案。
type ScriptRequest struct {
	VideoSubject       string `json:"video_subject"`
	VideoLanguage      string `json:"video_language"`
	ParagraphNumber    int    `json:"paragraph_number"`
	VideoScriptPrompt  string `json:"video_script_prompt"`
	CustomSystemPrompt string `json:"custom_system_prompt"`
}

// TermsRequest 仅生成素材关键词。
type TermsRequest struct {
	VideoSubject           string `json:"video_subject"`
	VideoScript            string `json:"video_script"`
	Amount                 int    `json:"amount"`
	MatchMaterialsToScript bool   `json:"match_materials_to_script"`
}

// SocialMetadataRequest 社交发布元数据。
type SocialMetadataRequest struct {
	VideoSubject string `json:"video_subject"`
	VideoScript  string `json:"video_script"`
	Language     string `json:"language"`
	Platform     string `json:"platform"`
}

// APIResponse 统一 JSON 外壳。
type APIResponse struct {
	Status  int    `json:"status"`
	Message string `json:"message,omitempty"`
	Data    any    `json:"data,omitempty"`
}

// VoicePreview WebUI 试听缓存。后台任务会再次核对文案和音色，过期则回退 TTS。
type VoicePreview struct {
	Script     string  `json:"script"`
	VoiceName  string  `json:"voice_name"`
	VoiceRate  float64 `json:"voice_rate"`
	VoiceVolume float64 `json:"voice_volume"`
	AudioFile  string  `json:"audio_file"`
	Duration   float64 `json:"duration"`
}

// LoomLoomVideoRequest 已确认报价的 LoomLoom 付费素材请求。
type LoomLoomVideoRequest struct {
	RunID             string `json:"run_id"`
	ListingVersionID  string `json:"listing_version_id"`
	ClientRequestID   string `json:"client_request_id"`
	APIToken          string `json:"api_token"`
	BaseURL           string `json:"base_url"`
}

// StartOptions 流水线可选入参，对应 Python start() 的关键字参数。
type StartOptions struct {
	StopAt          string
	AllowServerFile bool
	VoicePreview    *VoicePreview
	LoomLoomVideo   *LoomLoomVideoRequest
}

// TaskRecord 任务状态对外字段。
type TaskRecord struct {
	TaskID          string   `json:"task_id"`
	State           int      `json:"state"`
	Progress        int      `json:"progress"`
	Videos          []string `json:"videos,omitempty"`
	CombinedVideos  []string `json:"combined_videos,omitempty"`
	FailedStage     string   `json:"failed_stage,omitempty"`
	Error           string   `json:"error,omitempty"`
	Script          string   `json:"script,omitempty"`
	Terms           []string `json:"terms,omitempty"`
	AudioFile       string   `json:"audio_file,omitempty"`
	SubtitlePath    string   `json:"subtitle_path,omitempty"`
	Materials       []string `json:"materials,omitempty"`
	CrossPostState  string   `json:"cross_post_state,omitempty"`
	CrossPostError  string   `json:"cross_post_error,omitempty"`
	Warnings        []any    `json:"warnings,omitempty"`
	LoomLoomRunID   string   `json:"loomloom_run_id,omitempty"`
}

// MusicWarning 配乐降级时写入任务的警告。
func MusicWarning(code string, index int) map[string]any {
	return map[string]any{"code": code, "video_index": index}
}

// VideoMusicPrompt 读取当前配乐提示词：通用字段优先，Sonilo 兼容旧字段。
func (p VideoParams) VideoMusicPromptText() string {
	prompt := strings.TrimSpace(p.VideoMusicPrompt)
	if p.BGMType == "sonilo" && prompt == "" {
		return strings.TrimSpace(p.SoniloBGMPrompt)
	}
	return prompt
}

// IsBusy 判断任务是否仍在生成或发布，删除入口必须复用。
func (t *TaskRecord) IsBusy() bool {
	if t == nil {
		return false
	}
	if t.State == TaskStateProcessing {
		return true
	}
	switch t.CrossPostState {
	case CrossPostPending, CrossPostProcessing:
		return true
	}
	return false
}
